### DISCUSSION TITLE
Does it runs on a ESP32?

### DISCUSSION THREAD

**Discussion by Grabber** [nanomq/nanomq#455] [Q&A] [answered]
@JaylinYu, have you tried to run nanomq on an ESP32 (maybe the S3)?

**Comment by JaylinYu** [comment]
Hi Grabber, I haven't tried to port NanoMQ to ESP32-S3 before. However, I believe it is possible to run nanomq on it. The most important question is the flash memory, You will need to extend its memory by SPI since it only has 512 Kb, and contains BLE/WIFI stack which means it might now leave enough space for nanomq, and nanomq requires 300-400 kb as booting footprint. you can limit nanomq memory usage by stripping some unused protocol and debugging systems.

**Comment by Grabber** [comment]
It's sounding interesting to build an ESP32 tiny gateway with an embedded nanomq broker OR just use a tiny MQTT client (like the one from nanosdk) to forward messages coming to ESP32 directly to an external nanomq instance running anywhere else.

**Comment by Grabber** [comment]
https://arduinojson.org/v6/how-to/use-external-ram-on-esp32/

Probably by just enabling it the current code base will work:
```
CONFIG_SPIRAM_USE_MALLOC=y
```

**Comment by JaylinYu** [comment, marked as answer]
Bingo, If you don't need locality in your application, then build a client with MQTT SDK like nanosdk should certainly fit your goal better. It seems like to port POSIX pthread software from linux to esp32 may require a cross-compiling toolchain from https://docs.espressif.com/projects/esp-idf/en/stable/esp32s3/api-reference/system/pthread.html. But I am not sure, need to try it out.

**Comment by Grabber** [comment]
@JaylinYu 

Why there are 3 repos on nng? How are they different?

nanomq/NanoSDK
nanomq/nng 
nanomq/NanoNNG

**Comment by JaylinYu** [comment]
> 
nanomq/NanoSDK：  here is what we do with nanosdk(mqtt version of NNG), to make it independent so that user could find this sdk easily while our PR is not merged yet. And for us it is easier to manage the difference from upstream.

nanomq/nng ：this repo gonna be obsoleted. only for submitting PR to upstream. 

nanomq/NanoNNG ： it is what we do with broker stuff. This will be a diverge line from upstream, and should not be a concern from ordinary users.
